File: src/fireedge/src/client/features/OneApi/cluster.js
/* ------------------------------------------------------------------------- *
 * Copyright 2002-2024, OpenNebula Project, OpenNebula Systems               *
 *                                                                           *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may   *
 * not use this file except in compliance with the License. You may obtain   *
 * a copy of the License at                                                  *
 *                                                                           *
 * http://www.apache.org/licenses/LICENSE-2.0                                *
 *                                                                           *
 * Unless required by applicable law or agreed to in writing, software       *
 * distributed under the License is distributed on an "AS IS" BASIS,         *
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  *
 * See the License for the specific language governing permissions and       *
 * limitations under the License.                                            *
 * ------------------------------------------------------------------------- */
import { Actions, Commands } from 'server/utils/constants/commands/cluster'
import {
  Actions as ExtraActions,
  Commands as ExtraCommands,
} from 'server/routes/api/cluster/routes'

import {
  oneApi,
  ONE_RESOURCES,
  ONE_RESOURCES_POOL,
} from 'client/features/OneApi'
import { Cluster } from 'client/constants'
import {
  removeResourceOnPool,
  updateNameOnResource,
  updateResourceOnPool,
  updateTemplateOnResource,
} from 'client/features/OneApi/common'

const { CLUSTER, HOST, DATASTORE } = ONE_RESOURCES
const { CLUSTER_POOL, HOST_POOL, DATASTORE_POOL } = ONE_RESOURCES_POOL

const clusterApi = oneApi.injectEndpoints({
  endpoints: (builder) => ({
    getClusters: builder.query({
      /**
       * Retrieves information for all the clusters in the pool.
       *
       * @param {object} params - Request params
       * @param {string} [params.zone] - Zone from where to get the resources
       * @returns {Cluster[]} List of clusters
       * @throws Fails when response isn't code 200
       */
      query: (params) => {
        const name = Actions.CLUSTER_POOL_INFO
        const command = { name, ...Commands[name] }

        return { command, params }
      },
      transformResponse: (data) => [data?.CLUSTER_POOL?.CLUSTER ?? []].flat(),
      providesTags: (clusters) =>
        clusters
          ? [
              ...clusters.map(({ ID }) => ({
                type: CLUSTER_POOL,
                id: `${ID}`,
              })),
              CLUSTER_POOL,
            ]
          : [CLUSTER_POOL],
    }),
    getCluster: builder.query({
      /**
       * Retrieves information for the cluster.
       *
       * @param {object} params - Request params
       * @param {string} params.id - Cluster id
       * @param {boolean} [params.decrypt] - Optional flag to decrypt contained secrets, valid only for admin
       * @returns {Cluster} Get cluster identified by id
       * @throws Fails when response isn't code 200
       */
      query: (params) => {
        const name = Actions.CLUSTER_INFO
        const command = { name, ...Commands[name] }

        return { params, command }
      },
      transformResponse: (data) => data?.CLUSTER ?? {},
      providesTags: (_, __, { id }) => [{ type: CLUSTER, id }],
      async onQueryStarted({ id }, { dispatch, queryFulfilled }) {
        try {
          const { data: resourceFromQuery } = await queryFulfilled

          dispatch(
            clusterApi.util.updateQueryData(
              'getClusters',
              undefined,
              updateResourceOnPool({ id, resourceFromQuery })
            )
          )
        } catch {
          // if the query fails, we want to remove the resource from the pool
          dispatch(
            clusterApi.util.updateQueryData(
              'getClusters',
              undefined,
              removeResourceOnPool({ id })
            )
          )
        }
      },
    }),
    getClusterAdmin: builder.query({
      /**
       * Retrieve the information as serveradmin.
       *
       * @param {object} params - Request params
       * @param {string} params.id - Cluster id
       * @param {boolean} [params.decrypt] - Optional flag to decrypt contained secrets, valid only for admin
       * @returns {Cluster} Get cluster identified by id
       * @throws Fails when response isn't code 200
       */
      query: (params) => {
        const name = ExtraActions.CLUSTER_ADMINSHOW
        const command = { name, ...ExtraCommands[name] }

        return { params, command }
      },
      providesTags: (_, __, { id }) => [{ type: CLUSTER, id }],
    }),
    allocateCluster: builder.mutation({
      /**
       * Allocates a new cluster in OpenNebula.
       *
       * @param {object} params - Request params
       * @param {string} params.name - Name for the new cluster
       * @returns {number} The allocated cluster id
       * @throws Fails when response isn't code 200
       */
      query: (params) => {
        const name = Actions.CLUSTER_ALLOCATE
        const command = { name, ...Commands[name] }

        return { params, command }
      },
      invalidatesTags: [CLUSTER_POOL],
    }),
    removeCluster: builder.mutation({
      /**
       * Deletes the given cluster from the pool.
       *
       * @param {number|string} params - Cluster id
       * @returns {number} Cluster id
       * @throws Fails when response isn't code 200
       */
      query: (params) => {
        const name = Actions.CLUSTER_DELETE
        const command = { name, ...Commands[name] }

        return { params, command }
      },
      invalidatesTags: [CLUSTER_POOL],
      async onQueryStarted(params, { dispatch, queryFulfilled }) {
        try {
          const patchCluster = dispatch(
            clusterApi.util.updateQueryData(
              'getCluster',
              { id: params.id },
              updateNameOnResource(params)
            )
          )

          const patchClusters = dispatch(
            clusterApi.util.updateQueryData(
              'getClusters',
              undefined,
              updateNameOnResource(params)
            )
          )

          queryFulfilled.catch(() => {
            patchCluster.undo()
            patchClusters.undo()
          })
        } catch {}
      },
    }),
    updateCluster: builder.mutation({
      /**
       * Replaces the cluster template contents.
       *
       * @param {object} params - Request params
       * @param {number|string} params.id - Cluster id
       * @param {string} params.template - The new template contents
       * @param {0|1} params.replace
       * - Update type:
       * ``0``: Replace the whole template.
       * ``1``: Merge new template with the existing one.
       * @returns {number} Cluster id
       * @throws Fails when response isn't code 200
       */
      query: (params) => {
        const name = Actions.CLUSTER_UPDATE
        const command = { name, ...Commands[name] }

        return { params, command }
      },
      invalidatesTags: (_, __, { id }) => [{ type: CLUSTER, id }],
      async onQueryStarted(params, { dispatch, queryFulfilled }) {
        try {
          const patchCluster = dispatch(
            clusterApi.util.updateQueryData(
              'getCluster',
              { id: params.id },
              updateTemplateOnResource(params)
            )
          )

          const patchClusters = dispatch(
            clusterApi.util.updateQueryData(
              'getClusters',
              undefined,
              updateTemplateOnResource(params)
            )
          )

          queryFulfilled.catch(() => {
            patchCluster.undo()
            patchClusters.undo()
          })
        } catch {}
      },
    }),
    addHostToCluster: builder.mutation({
      /**
       * Adds a host to the given cluster.
       *
       * @param {object} params - Request params
       * @param {number|string} params.id - The cluster id
       * @param {string} params.host - The host id
       * @returns {number} Cluster id
       * @throws Fails when response isn't code 200
       */
      query: (params) => {
        const name = Actions.CLUSTER_ADDHOST
        const command = { name, ...Commands[name] }

        return { params, command }
      },
      invalidatesTags: (_, __, { id, host }) => [
        { type: CLUSTER, id },
        { type: HOST, id: host },
        CLUSTER_POOL,
        HOST_POOL,
      ],
    }),
    removeHostFromCluster: builder.mutation({
      /**
       * Removes a host from the given cluster.
       *
       * @param {object} params - Request params
       * @param {number|string} params.id - The cluster id
       * @param {string} params.host - The host id
       * @returns {number} Cluster id
       * @throws Fails when response isn't code 200
       */
      query: (params) => {
        const name = Actions.CLUSTER_DELHOST
        const command = { name, ...Commands[name] }

        return { params, command }
      },
      invalidatesTags: (_, __, { id }) => [{ type: CLUSTER, id }, CLUSTER_POOL],
    }),
    addDatastoreToCluster: builder.mutation({
      /**
       * Adds a datastore to the given cluster.
       *
       * @param {object} params - Request params
       * @param {number|string} params.id - The cluster id
       * @param {string} params.datastore - The datastore id
       * @returns {number} Cluster id
       * @throws Fails when response isn't code 200
       */
      query: (params) => {
        const name = Actions.CLUSTER_ADDDATASTORE
        const command = { name, ...Commands[name] }

        return { params, command }
      },
      invalidatesTags: (_, __, { id, datastore }) => [
        { type: CLUSTER, id },
        CLUSTER_POOL,
        DATASTORE_POOL,
        { type: DATASTORE, id: datastore },
      ],
    }),
    removeDatastoreFromCluster: builder.mutation({
      /**
       * Removes a datastore from the given cluster.
       *
       * @param {object} params - Request params
       * @param {number|string} params.id - The cluster id
       * @param {string} params.datastore - The datastore id
       * @returns {number} Cluster id
       * @throws Fails when response isn't code 200
       */
      query: (params) => {
        const name = Actions.CLUSTER_DELDATASTORE
        const command = { name, ...Commands[name] }

        return { params, command }
      },
      invalidatesTags: (_, __, { id, datastore }) => [
        { type: CLUSTER, id },
        CLUSTER_POOL,
        DATASTORE_POOL,
        { type: DATASTORE, id: datastore },
      ],
    }),
    addNetworkToCluster: builder.mutation({
      /**
       * Adds a vnet to the given cluster.
       *
       * @param {object} params - Request params
       * @param {number|string} params.id - The cluster id
       * @param {string} params.vnet - The vnet id
       * @returns {number} Cluster id
       * @throws Fails when response isn't code 200
       */
      query: (params) => {
        const name = Actions.CLUSTER_ADDVNET
        const command = { name, ...Commands[name] }

        return { params, command }
      },
      invalidatesTags: (_, __, { id }) => [{ type: CLUSTER, id }, CLUSTER_POOL],
    }),
    removeNetworkFromCluster: builder.mutation({
      /**
       * Removes a vnet from the given cluster.
       *
       * @param {object} params - Request params
       * @param {number|string} params.id - The cluster id
       * @param {string} params.vnet - The vnet id
       * @returns {number} Cluster id
       * @throws Fails when response isn't code 200
       */
      query: (params) => {
        const name = Actions.CLUSTER_DELVNET
        const command = { name, ...Commands[name] }

        return { params, command }
      },
      invalidatesTags: (_, __, { id }) => [{ type: CLUSTER, id }, CLUSTER_POOL],
    }),
    renameCluster: builder.mutation({
      /**
       * Renames a cluster.
       *
       * @param {object} params - Request parameters
       * @param {string|number} params.id - Cluster id
       * @param {string} params.name - The new name
       * @returns {number} Cluster id
       * @throws Fails when response isn't code 200
       */
      query: (params) => {
        const name = Actions.CLUSTER_RENAME
        const command = { name, ...Commands[name] }

        return { params, command }
      },
      invalidatesTags: (_, __, { id }) => [{ type: CLUSTER, id }, CLUSTER_POOL],
    }),
  }),
})

export const {
  // Queries
  useGetClustersQuery,
  useLazyGetClustersQuery,
  useGetClusterQuery,
  useLazyGetClusterQuery,
  useGetClusterAdminQuery,
  useLazyGetClusterAdminQuery,

  // Mutations
  useAllocateClusterMutation,
  useRemoveClusterMutation,
  useUpdateClusterMutation,
  useAddHostToClusterMutation,
  useRemoveHostFromClusterMutation,
  useAddDatastoreToClusterMutation,
  useRemoveDatastoreFromClusterMutation,
  useAddNetworkToClusterMutation,
  useRemoveNetworkFromClusterMutation,
  useRenameClusterMutation,
} = clusterApi

export default clusterApi

File: src/ods/lib/helpers/one/cluster.rb
# -------------------------------------------------------------------------- #
# Copyright 2002-2026, OpenNebula Project, OpenNebula Systems                #
#                                                                            #
# Licensed under the Apache License, Version 2.0 (the "License"); you may    #
# not use this file except in compliance with the License. You may obtain    #
# a copy of the License at                                                   #
#                                                                            #
# http://www.apache.org/licenses/LICENSE-2.0                                 #
#                                                                            #
# Unless required by applicable law or agreed to in writing, software        #
# distributed under the License is distributed on an "AS IS" BASIS,          #
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.   #
# See the License for the specific language governing permissions and        #
# limitations under the License.                                             #
#--------------------------------------------------------------------------- #

module OpenNebula

    module DocumentServer

        # Defines methods to manage resources in OpenNebula using the OCA API
        module OneHelper

            # Defines methods to manage OpenNebula Clusters.
            module Cluster

                RESOURCE_TYPE = 'cluster'
                WAIT_DELETE   = false
                POOL_CLASS    = OpenNebula::ClusterPool

                # Creates a cluster and optionally appends template attributes.
                # @param client [OpenNebula::Client] OpenNebula client.
                # @param name [String] Cluster name.
                # @param template [Hash] Initial template attributes.
                # @param extra_template [Hash] Additional template attributes.
                # @return [OpenNebula::Cluster, OpenNebula::Error] created cluster or an API error.
                def self.create(client, name:, template: {}, extra_template: {}, **_options)
                    return OpenNebula::Error.new(
                        'Cluster name cannot be empty', OpenNebula::Error::EACTION
                    ) if name.to_s.empty?

                    cluster = OpenNebula::Cluster.new(OpenNebula::Cluster.build_xml, client)
                    rc      = cluster.allocate(name)
                    return rc if OpenNebula.is_error?(rc)

                    content = (template || {}).merge(extra_template || {})

                    unless content.empty?
                        content = Hash.to_raw(content)
                        return content if OpenNebula.is_error?(content)

                        rc = cluster.update(content, true)
                        return rc if OpenNebula.is_error?(rc)
                    end

                    rc = cluster.info
                    return rc if OpenNebula.is_error?(rc)

                    cluster
                end

                # Retrieves a cluster with its current information.
                # @param client [OpenNebula::Client] OpenNebula client.
                # @param cluster_id [Integer] Cluster ID.
                # @return [OpenNebula::Cluster, OpenNebula::Error] cluster or an API error.
                def self.get(client, cluster_id)
                    return OpenNebula::Error.new(
                        'Cluster ID cannot be nil', OpenNebula::Error::EACTION
                    ) if cluster_id.nil?

                    cluster = OpenNebula::Cluster.new_with_id(cluster_id, client)

                    rc = cluster.info
                    return rc if OpenNebula.is_error?(rc)

                    cluster
                end

                # Returns a cluster body using symbolized keys.
                # @param client [OpenNebula::Client] OpenNebula client.
                # @param cluster_id [Integer] Cluster ID.
                # @param downcase [Boolean] Whether to downcase keys.
                # @return [Hash, OpenNebula::Error] cluster body or an API error.
                def self.body(client, cluster_id, downcase: true)
                    cluster = get(client, cluster_id)
                    return cluster if OpenNebula.is_error?(cluster)

                    body = cluster.to_hash['CLUSTER']

                    return OpenNebula::Error.new(
                        "Cannot retrieve OpenNebula cluster body for resource '#{cluster_id}'",
                        OpenNebula::Error::EACTION
                    ) unless body

                    body.deep_symbolize_keys(:downcase => downcase)
                end

                # Checks whether a cluster with a name exists.
                # @param client [OpenNebula::Client] OpenNebula client.
                # @param name [String] Cluster name.
                # @return [Boolean, OpenNebula::Error] existence result or an API error.
                def self.exists?(client, name)
                    cluster = find(client, name)
                    return cluster if OpenNebula.is_error?(cluster)

                    !cluster.nil?
                end

                # Checks whether a cluster ID exists in the pool.
                # @param client [OpenNebula::Client] OpenNebula client.
                # @param cluster_id [Integer] Cluster ID.
                # @return [Boolean, OpenNebula::Error] existence result or an API error.
                def self.exists_id?(client, cluster_id)
                    cluster_pool = OpenNebula::ClusterPool.new(client)

                    rc = cluster_pool.info
                    return rc if OpenNebula.is_error?(rc)

                    cluster_pool.any? {|cluster| cluster.id.to_i == cluster_id.to_i }
                end

                # Finds a cluster by name.
                # @param client [OpenNebula::Client] OpenNebula client.
                # @param name [String] Cluster name.
                # @return [OpenNebula::Cluster, nil, OpenNebula::Error]
                def self.find(client, name)
                    cluster_pool = OpenNebula::ClusterPool.new(client)

                    rc = cluster_pool.info
                    return rc if OpenNebula.is_error?(rc)

                    cluster = cluster_pool.find {|item| item.name == name }
                    return if cluster.nil?

                    rc = cluster.info
                    return rc if OpenNebula.is_error?(rc)

                    cluster
                end

                # Deletes a cluster, optionally waiting for its removal.
                # @param client [OpenNebula::Client] OpenNebula client.
                # @param cluster_id [Integer] Cluster ID.
                # @param wait [Boolean] Whether to wait for deletion.
                # @return [true, OpenNebula::Error] success or an API error.
                def self.delete(client, cluster_id, wait: false)
                    cluster = get(client, cluster_id)
                    return cluster if OpenNebula.is_error?(cluster)

                    rc = cluster.delete
                    return rc if OpenNebula.is_error?(rc)
                    return true unless wait

                    Resource.wait_until_deleted(cluster)
                end

                # Lists the resources associated with a cluster, grouped by type.
                # @param client [OpenNebula::Client] OpenNebula client.
                # @param cluster_id [Integer] Cluster ID.
                # @return [Hash, OpenNebula::Error] non-empty resource IDs by type or an API error.
                def self.associated_objects(client, cluster_id)
                    cluster = get(client, cluster_id)
                    return cluster if OpenNebula.is_error?(cluster)

                    body = cluster.to_hash['CLUSTER'] || {}

                    {
                        'host'      => Array(body.dig('HOSTS', 'ID')),
                        'datastore' => Array(body.dig('DATASTORES', 'ID')),
                        'network'   => Array(body.dig('VNETS', 'ID'))
                    }.reject {|_, ids| ids.empty? }
                end

            end

        end

    end

end
